lib/kitchen/provisioner/chef_zero.rb

```diff
@@ -195,6 +207,23 @@ def prepare_chef_client_zero_rb
         FileUtils.cp(source, File.join(sandbox_path, "chef-client-zero.rb"))
       end
 
+      HANDLER_CODE = <<-EOH
+# Handler to kill the run if any resource is updated
+class UpdatedResources < ::Chef::Handler
+  def report
+    if updated_resources.size > 0
+      #exit 203 # if chef handler code was not catching Exception
+      puts "Chef run should reach a converged state. Resources updated in a second chef-client run:"
+      updated_resources.each do |r|
+        puts r
+      end
+      Process.kill('KILL', Process.pid)
```

Comment:
handler code catches all exceptions (https://github.com/chef/chef/blob/master/lib/chef/handler.rb#L215) which prevent to use a classic exception raise to end the run.

I agree this is a hack, I'd be happy to find another solution.